From 6e8a642a1395b337e7254110eacffec64d67cf2d Mon Sep 17 00:00:00 2001
From: Mika Westerberg <mika.westerberg@linux.intel.com>
Date: Mon, 19 Sep 2016 18:17:45 +0300
Subject: [PATCH 8/8] spi: pxa2xx: Add support for GPIO chip selects provided
 by the SPI core

The SPI core parses and requests GPIO chip selects from firmware provided
hardware configuration tables (like ACPI) and assigns these for each SPI
slave device by using spi->cs_gpiod field. If this field is not NULL it can
be used to toggle GPIO chip select for the device.

Signed-off-by: Mika Westerberg <mika.westerberg@linux.intel.com>
---
 drivers/spi/spi-pxa2xx.c | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/drivers/spi/spi-pxa2xx.c b/drivers/spi/spi-pxa2xx.c
index 87150a1049bd..6814b400e1de 100644
--- a/drivers/spi/spi-pxa2xx.c
+++ b/drivers/spi/spi-pxa2xx.c
@@ -1178,7 +1178,16 @@ static int setup_cs(struct spi_device *spi, struct chip_data *chip,
 {
 	int err = 0;
 
-	if (chip == NULL || chip_info == NULL)
+	if (chip == NULL)
+		return 0;
+
+	if (spi->cs_gpiod) {
+		chip->gpio_cs = desc_to_gpio(spi->cs_gpiod);
+		chip->gpio_cs_inverted = spi->mode & SPI_CS_HIGH;
+		return 0;
+	}
+
+	if (chip_info == NULL)
 		return 0;
 
 	/* NOTE: setup() can be called multiple times, possibly with
@@ -1352,7 +1361,8 @@ static void cleanup(struct spi_device *spi)
 	if (!chip)
 		return;
 
-	if (drv_data->ssp_type != CE4100_SSP && gpio_is_valid(chip->gpio_cs))
+	if (drv_data->ssp_type != CE4100_SSP && !spi->cs_gpiod &&
+	    gpio_is_valid(chip->gpio_cs))
 		gpio_free(chip->gpio_cs);
 
 	kfree(chip);
-- 
2.9.3

